### ISSUE TITLE
SEC Cleanup

### ISSUE THREAD

**Issue by krpeacock** [dfinity/invoice-canister#29]
7 additional notes were provided from the security audit that have minimal risk and are trivial to clean up.

- redundant argument in verify_invoice: The caller is no longer needed in ICPVerifyInvoiceArgs. It may be safer to remove it to reduce the risk of confusing the caller with the creator or so. 
- In verify_invoice the from_subaccount is re-computed. Would it be better to add this information to the invoice when creating it? IMO That would be safer / easier to review since the info would be generated in a single place. 
- verify_invoice still has a TODO that should be removed. 
- in refund_invoice let replaced = invoices.put(i.id, updatedInvoice) is irritating since put does not return anything according to the docs. Also, in refund_invoice this variable is never used. 
- get_invoice returns invoice by default: If the switch statement does not return, the method returns the invoice by default (last statement in get_invoice). 
- get_invoice permissions: IIUC one always needs to put the buyer principal in the canGet permissions. Is this intended and does one always know the buyer when creating the invoice? 
- unused code: defaultSubAccount is never used and should be deleted

**Comment by atengberg** [comment]
Just a note, generating the payment address when the invoice is created and including it in the invoice's metadata seems like the cleanest way to resolve the second bullet point--I would assume it should only be visible to the invoice creator's principal (and by extension to the buyer if requested or returned to them when the invoice is originally created). 

As far as the buyer (I take to mean the customer) principal is concerned, is the buyer's principal necessary for any part of the actual access control / permissions of the invoice? Since any of the buyer's requests are done through the seller any of those calls would actually be come from the seller's principal (as the buyer doesn't directly interact with the invoice canister itself).

**Comment by krpeacock** [comment]
I agree - it can be covered under the same `read` permissions as the rest of the metadata

**Comment by AshtonEngberg** [comment]
One other question I had is that currently the E2E tests do not use directly use the ledger canister: funds are initially deposited to the subaccount of one of the identities used in the E2E tests (the balance holder), and then passed back and forth between it and the other identity created as invoices are created, paid and verified; that is, all the funds are in the custody of the invoice canister the whole time. Not that this makes any difference (as opposed to transferring funds out of the invoice canister's custody and verifying the balance by querying the ledger canister directly) and indeed simplifies the testing, but with the introduction of the ICRC-1 token integration, should the tests be refactored to directly call on the ledger canister(s)? 

This introduces slightly more work, but enumerates using the ledger canister(s)'s in conjunction with the invoice canister, which might fit the theme of the examples repo better. Otherwise I can simplify the code base as the ledger canister(s) declarations are not needed in the E2E testing.

**Comment by AshtonEngberg** [comment]
One question that has come up is with regard to the method get_caller_identifier in the main.mo that allows any caller to get the address of the corresponding to the principal subaccount that is passed in as the principal in the GetAccountIdentifierArgs. Is this not a security concern? 

Mitigation could be making this a shared function so that only the caller can get their own address and dropping the principal argument: if this functionality is also needed for administrative purposes, that principal argument can be made opt and then only allowed to return the corresponding address if the caller belongs to the administrator set (the same set who can add and remove principals from the creator allow list).

Since this isn't part of the bounty, but seems part of the intent of the bounty, is it alright if I make this change in the process of completing the other bounty tasks?

**Comment by atengberg** [comment]
Any thoughts on "In verify_invoice the from_subaccount is re-computed" ? 

Initially I agreed with this, but actually computing the subaccount is direct and only requires the id and creator principal which are immutable fields of a given invoice and thus always available when this subaccount would be used for from_subaccount.
